cmd/bashbrew/cmd-build.go

```diff
@@ -87,51 +89,50 @@ func cmdBuild(c *cli.Context) error {
-
 					switch builder := entry.ArchBuilder(arch); builder {
-					case "classic", "":
+					case "buildkit", "classic", "":
 						var platform string
 						if fromScratch {
 							platform = ociArch.String()
 						}
-						err = dockerBuild(cacheTag, entry.ArchFile(arch), archive, platform)
+
+						archive, err := gitArchive(commit, entry.ArchDirectory(arch))
 						if err != nil {
-							return cli.NewMultiError(fmt.Errorf(`failed building %q (tags %q)`, r.RepoName, entry.TagsString()), err)
+							return cli.NewMultiError(fmt.Errorf(`failed generating git archive for %q (tags %q)`, r.RepoName, entry.TagsString()), err)
 						}
-					case "buildkit":
-						var platform string
-						if fromScratch {
-							platform = ociArch.String()
+						defer archive.Close()
+
+						if builder == "buildkit" {
+							err = dockerBuildxBuild(tags, entry.ArchFile(arch), archive, platform)
+						} else {
+							// TODO use "meta.StageNames" to do "docker build --target" so we can tag intermediate stages too for cache (streaming "git archive" directly to "docker build" makes that a little hard to accomplish without re-streaming)
+							err = dockerBuild(tags, entry.ArchFile(arch), archive, platform)
 						}
-						err = dockerBuildxBuild(cacheTag, entry.ArchFile(arch), archive, platform)
 						if err != nil {
 							return cli.NewMultiError(fmt.Errorf(`failed building %q (tags %q)`, r.RepoName, entry.TagsString()), err)
 						}
+
+						archive.Close() // be sure this happens sooner rather than later (defer might take a while, and we want to reap zombies more aggressively)
+
+					case "oci-import":
+						err := ociImportBuild(tags, commit, entry.ArchDirectory(arch), entry.ArchFile(arch))
+						if err != nil {
+							return cli.NewMultiError(fmt.Errorf(`failed oci-import build of %q (tags %q)`, r.RepoName, entry.TagsString()), err)
+						}
+
+						fmt.Printf("Importing %s into Docker\n", r.EntryIdentifier(entry))
+						err = ociImportDockerLoad(imageTags)
+						if err != nil {
+							return cli.NewMultiError(fmt.Errorf(`failed oci-import into Docker of %q (tags %q)`, r.RepoName, entry.TagsString()), err)
+						}
+
 					default:
 						return cli.NewMultiError(fmt.Errorf(`unknown builder %q`, builder))
 					}
-					archive.Close() // be sure this happens sooner rather than later (defer might take a while, and we want to reap zombies more aggressively)
 				}
 			} else {
 				fmt.Printf("Using %s (%s)\n", cacheTag, r.EntryIdentifier(entry))
 			}
-
-			for _, tag := range r.Tags(namespace, uniq, entry) {
-				fmt.Printf("Tagging %s\n", tag)
-				if !dryRun {
-					err := dockerTag(cacheTag, tag)
-					if err != nil {
-						return cli.NewMultiError(fmt.Errorf(`failed tagging %q as %q`, cacheTag, tag), err)
-					}
-				}
-			}
```

Comment:
Making this part of `docker build` has the side effect that if we skip `docker build` (because `bashbrew/cache:xxx` exists and is up-to-date) it might not get _all_ the tags it's supposed to have. :weary:

Reply:
I see three primary options here:

1. revert this part of the change and go back to `docker build` + `docker tag` in a loop
2. rely on `docker build` cache exclusively
    - noting that we added this indirection specifically because that cache was horrifyingly slow at the large scale our builders typically run
3. when our bashbrew cache exists already, only then do the `docker tag` loop
    - on Windows, it would be faster and more reliable to `docker inspect` each one before we `docker tag` it (and `docker tag` often fails, so it'd be useful to have it auto-retry once or twice, but only on Windows :sob:)

Reply:
I guess we also could:

4. do a basic `docker build` that we pass just `FROM bashbrew/cache:xxx` to abuse `docker build` as a multi-`docker tag` :sweat_smile:

Reply:
(To expound, a legit case where this gets hit is that we push a build, it works, it finishes, and then we add a new tag or change a tag and this now will fail pushing.)

Reply:
Assumption: `docker build -t '' -t '' ...` is faster than multiple `docker tag`, especially for Windows.

I think using `docker build` to get multiple tags at once is nice. It could speed up builds (and maybe help windows fail less?). I'm not sure about using a single line `FROM` Dockerfile; it feels bad, like something that could be inadvertently broken by buildkit in the future.

As far as number two, we know that classic `docker build` cache is slow, so we can revisit this possibility once we use buildkit by default and we can try relying on its cache (I'll start testing it).

On windows we should do the inspect to see if that specific tag is already correct. :see_no_evil: 

So I guess my conclusion is to probably use the hacky single line `FROM` Dockerfile to do a multi-tag.  If it breaks in the future we can hopefully just rely on buildkit build cache (or maybe convince `docker tag` to take multiple targets).